### PR TITLE
Clang warnings cleanup

### PR DESCRIPTION
**Pull request by cbezault**
I noticed that our code emits a fair number of warnings with `clang-cl` when building the tests. Previously we were not seeing these warnings because of the way `clang` treats "system headers".

This change causes `clang-cl` to emit warnings on our headers and fixes up the existing warnings.

I also resolved Microsoft-internal VSO-609129 "<cvt/sjis_0208> appears to have impossible condition" which was being warned about by clang.

### CODE REVIEW

File: tests/std/tests/P0067R5_charconv/env.lst [hunk 1]
```diff
@@ -1,6 +1,34 @@
...
+PM_CL="/clr /MD /std:c++17"
+PM_CL="/clr /MDd /std:c++17"
+PM_CL="/BE /c /EHsc /MD /std:c++latest"
+PM_CL="/BE /c /EHsc /MDd /std:c++17 /permissive-"
+PM_CL="/BE /c /EHsc /MTd /std:c++latest /permissive-"
+PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing /EHsc /MD /std:c++latest"
+PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing /EHsc /MDd /std:c++17"
+PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing /EHsc /MTd /std:c++latest /fp:strict"
 RUNALL_CROSSLIST
 PM_CL="/O2 /wd4793"
```
Comment (CaseyCarter) [inline tests/std/tests/P0067R5_charconv/env.lst:34]: Can't we simply merge this with line 6 and drop line 33?

File: tests/utils/stl/test/config.py [hunk 2]
```diff
@@ -296,22 +259,10 @@ def configure_path_env_var(self):
         self.stl_path_env_var = stl_path_env_var
 
     def configure_test_env(self):
-        stl_test_env = self.get_lit_conf('stl_test_env', None)
-
-        if stl_test_env is None:
-            stl_test_env = self.config.environment
-
-        if self.stl_lib_env_var is None:
```
Comment (CaseyCarter) [inline tests/utils/stl/test/config.py:304]: Oops. How'd we miss this is `lib` instead of `inc`? (No change requested,)

File: stl/inc/random [hunk 3]
```diff
@@ -736,7 +736,7 @@ protected:
         this->_Idx = _Rx;
     }
 
-    void _Setx(unsigned int _Ix, _Ty _Xis, _Ty _Xir) { // update _Ax[_Ix] and _Carry
+    void _Setx(unsigned size_t _Ix, _Ty _Xis, _Ty _Xir) { // update _Ax[_Ix] and _Carry
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/random]: Looks like this was resolved before I finished my review.

File: stl/inc/type_traits [hunk 4]
```diff
@@ -823,6 +823,12 @@ struct is_trivially_destructible : bool_constant<__is_trivially_destructible(_Ty
 template <class _Ty>
 _INLINE_VAR constexpr bool is_trivially_destructible_v = __is_trivially_destructible(_Ty);
 
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-volatile"
+#endif
```
[the remark this answers is not in the document]
Comment (cbezault) [reply stl/inc/type_traits:830]: Yup that was a mistake on my end. Might have gotten introduced when cherry-picking commits.

[the remark this answers is not in the document]
Comment (cbezault) [reply stl/inc/type_traits:830]: There was test code triggering it here and I just threw the suppression here. That's wrong.

File: stl/inc/xutility [hunk 5]
```diff
@@ -4941,7 +4941,8 @@ _NODISCARD _CONSTEXPR20 _InIt _Find_unchecked1(_InIt _First, const _InIt _Last,
 
 #ifdef __cpp_lib_is_constant_evaluated
     if (_STD is_constant_evaluated()) {
-        return _Find_unchecked1(_First, _Last, _Val, false_type{});
+        using _Elem = remove_pointer_t<_InIt>;
+        return _Find_unchecked1(_First, _Last, static_cast<_Elem>(_Val), false_type{});
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xutility:4945]: My mistake: I didn't notice that the metaprogramming guarantees that both `_Ty` and the pointee type of `_InIt` are integer types, so the comparisons aren't observable.

File: stl/inc/exception [hunk 6]
```diff
@@ -171,7 +171,8 @@ inline terminate_handler __CRTDECL set_terminate(terminate_handler) noexcept { /
     return nullptr;
 }
 
-inline void __CRTDECL terminate() noexcept { // handle exception termination
+[[noreturn]] inline void __CRTDECL terminate() noexcept { // handle exception termination
+    _CSTD abort();
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/exception:175]: I noticed this a while ago and thought I had recorded a TODO, but didn't. (I'm migrating the TODO list to `enhancement` issues - 5 pages to go!) I agree that we should file an issue for this.

File: stl/inc/yvals.h [hunk 7]
```diff
@@ -184,6 +184,19 @@ _STL_DISABLE_CLANG_WARNINGS
         _CRT_SECURE_INVALID_PARAMETER(mesg); \
     } while (false)
 
+#ifdef __clang__
+#define _STL_VERIFY(cond, mesg)                                                            \
+    _Pragma("clang diagnostic push") _Pragma("clang diagnostic ignored \"-Wassume\"") do { \
```
[the remark this answers is not in the document]
Comment (cbezault) [reply stl/inc/yvals.h:175]: It's supposed to warn about discarded side-effects in the argument to `__asume`. However, I just re-ran the tests on Billy's computer without this change and can't reproduce the warnings I was getting which made me add this so I'm just going to drop the change.

File: stl/inc/type_traits [hunk 8]
```diff
@@ -2360,5 +2366,10 @@ _STD_END
 _STL_RESTORE_CLANG_WARNINGS
 #pragma warning(pop)
 #pragma pack(pop)
+
+#ifdef __clang__
+#pragma clang diagnostic pop
```
Comment (CaseyCarter) [inline stl/inc/type_traits:2371]: The MSVC pop which corresponds to this Clang pop is on line 1679. Could we move this one closer?

File: stl/inc/yvals.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (cbezault) [reply stl/inc/yvals.h]: Turns out it's still there: https://dev.azure.com/vclibs/STL/_build/results?buildId=2784&view=ms.vss-test-web.build-test-results-tab&runId=1004806&resultId=108940&paneView=debug

File: stl/inc/cvt/sjis_0208 [hunk 9]
```diff
@@ -57,7 +57,7 @@ namespace stdext {
                         // convert 2-byte code
                         unsigned char _By2 = static_cast<unsigned char>(*++_Mid1);
 
-                        if ((_By2 < 0x40 || 0x7e < _By2) && (_By2 < 0x80 && 0xfc < _By2)) { // TRANSITION, VSO-609129
+                        if (_By2 < 0x40 || _By2 == 0x7f || _By2 > 0xfc) {
```
Comment (StephanTLavavej) [inline stl/inc/cvt/sjis_0208:60]: Thanks for fixing this! (Microsoft-internal VSO-609129 "`<cvt/sjis_0208>` appears to have impossible condition".)

While making this behavioral change as part of a warning cleanup is fine (especially if Clang was warning about a condition that can never be true), it should be specifically called out in the checkin notes.

File: tests/utils/stl/test/config.py [hunk 10]
```diff
@@ -42,7 +42,6 @@ def __init__(self, lit_config, config):
         self.long_tests = None
         self.msvc_toolset_libs_root = None
         self.stl_build_root = None
-        self.stl_inc_env_var = None
         self.stl_lib_env_var = None
```
Comment (StephanTLavavej) [inline tests/utils/stl/test/config.py:45]: Is `self.stl_lib_env_var` unused now?

File: tests/utils/stl/test/config.py [hunk 11]
```diff
@@ -296,22 +259,10 @@ def configure_path_env_var(self):
...
-
-        if self.stl_lib_env_var is None:
-            self.configure_lib_env_var()
-
         if self.stl_path_env_var is None:
             self.configure_path_env_var()
 
-        stl_test_env['INCLUDE'] = self.stl_inc_env_var
-        stl_test_env['LIB'] = self.stl_lib_env_var
+        stl_test_env = {}
```
Comment (StephanTLavavej) [inline tests/utils/stl/test/config.py:265]: How is the freshly-built STL getting picked up by the test infrastructure now? That is, if I replace `<vector>` with `#error WOOF`, will it still cause the tests to fail?

File: stl/inc/yvals.h [hunk 12]
```diff
@@ -170,6 +170,19 @@ _STL_DISABLE_CLANG_WARNINGS
...
```
[the remark this answers is not in the document]
Comment (cbezault) [reply stl/inc/yvals.h:175]: clang-format made me do it.

File: tests/utils/stl/test/config.py (hunk 11 above)
Comment (cbezault) [reply tests/utils/stl/test/config.py:265]: I hard-code `/I`s and `/LIBPATH`s.

Comment (StephanTLavavej) [reply tests/utils/stl/test/config.py:265]: Oh, that was already being done elsewhere, got it.

File: stl/inc/yvals.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/yvals.h:175]: Ugh. Ok to leave as-is then.

File: tests/utils/stl/test/config.py (hunk 11 above)
Comment (cbezault) [reply tests/utils/stl/test/config.py:265]: And `PATH` is still getting set to pick up the dlls.

File: stl/inc/type_traits [hunk 13]
```diff
@@ -1665,12 +1665,24 @@ struct _Invoker_ret<_Unforced, false> { // selected for _Rx being _Unforced
 #pragma warning(disable : 4244) // 'argument': conversion from '_From' to '_To', possible loss of data
 #pragma warning(disable : 4365) // 'argument': conversion from '_From' to '_To', signed/unsigned mismatch (/Wall)
 #pragma warning(disable : 5215) // '%s' a function parameter with a volatile qualified type is deprecated in C++20
+
+#ifdef __clang__
+#pragma clang diagnostic push
```
Comment (barcharcraz) [inline stl/inc/type_traits:1670]: do we want a transition tag here for when we remove volatile from our headers?

File: tests/tr1/tests/cvt/sjis_0208/test.cpp [hunk 14]
```diff
@@ -3,9 +3,97 @@
...
+    // First set of two byte characters with odd first bytes
+    for (unsigned long ch = 0x8100; ch <= 0xa000; ch += 0x0200) {
+        for (unsigned long offset = 0x40; offset < 0x7f; ++offset) {
+            string buf = myconv.to_bytes((MYWCHAR)(ch + offset));
+            if (0 < buf.size() && (ch + offset) != static_cast<unsigned long>(WEOF)) {
+                mystring.insert(mystring.end(), (MYWCHAR)(ch + offset));
+            }
+        }
+
+        for (unsigned long offset = 0x8A; offset < 0x9f; ++offset) {
```
Comment (barcharcraz) [inline tests/tr1/tests/cvt/sjis_0208/test.cpp:51]: this stuff looks like unrelated changes?

Comment (StephanTLavavej) [reply tests/tr1/tests/cvt/sjis_0208/test.cpp:51]: This tests the fix for VSO-609129 "`<cvt/sjis_0208>` appears to have impossible condition", which is sufficient to override our usual practice of "no new test coverage in tr1".

File: stl/inc/type_traits (hunk 13 above)
Comment (StephanTLavavej) [reply stl/inc/type_traits:1670]: We generally use `TRANSITION` to mark code that should be changed when compiler bugs are fixed, or compiler features are implemented. Here, I don't believe that we could remove these warning suppressions unless and until the Standard changes to forbid `volatile` inputs into this machinery (e.g. nobody should be able to ask whether something is nothrow implicitly convertible to `volatile X`). That might happen in C++23, or maybe never. In any event, it would be changed by a paper (or possibly an LWG issue), which we already have systems for tracking, so I don't believe that `TRANSITION` comments would achieve anything here.

File: tests/tr1/tests/cvt/sjis_0208/test.cpp (hunk 14 above)
Comment (cbezault) [reply tests/tr1/tests/cvt/sjis_0208/test.cpp:51]: Yeah like Stephan said this is to test the cvt fix. Most of the cvt tests use a generic character generator to generate the full set of legal characters to test against but sjis_0208 is a little weird so I decided to just write a specialized character generator.

File: tests/tr1/tests/cvt/sjis_0208/test.cpp [hunk 15]
```diff
@@ -3,9 +3,97 @@
...
+using Myostream  = basic_ostream<MYWCHAR>;
+using Mystring   = basic_string<MYWCHAR>;
+using Mysconvert = stdext::cvt::wstring_convert<MYNAME, MYWCHAR>;
+
+Mystring mymake() {
+    Mystring mystring;
+    Mysconvert myconv("");
+
+    // All ASCII and modified ASCII characters single byte characters
+    for (unsigned long ch = 0; ch <= 0x7f; ++ch) {
```
Comment (cbezault) [inline tests/tr1/tests/cvt/sjis_0208/test.cpp:27]: Hm, I guess this introduces a ton of unnecessary `nul` characters to be tested but still works.

**Comment by StephanTLavavej** [comment]
Thanks for fixing all of these warnings, *and* the cvt bugfix, **and especially** enhancing our test coverage so we remain clean in the future! I had dreamed of this but didn't know how to achieve it in the old infrastructure (as `-Wsystem-headers` affects UCRT/WinSDK headers too). Amazing work.
